cdc/cdc.cc

```diff
@@ -0,0 +1,197 @@
+    auto& mm = service::get_local_migration_manager();
+    return mm.announce_column_family_drop(ks_name, log_name(table_name), false);
+}
+
+static future<> remove_desc(const sstring& ks_name, const sstring& table_name) {
+    auto& mm = service::get_local_migration_manager();
+    return mm.announce_column_family_drop(ks_name, desc_name(table_name), false);
+}
+
+future<> remove(const sstring& ks_name, const sstring& table_name) {
+    return when_all(remove_log(ks_name, table_name), remove_desc(ks_name, table_name)).discard_result();
+}
+
+static future<> setup_log(const schema& s) {
+    schema_builder b(s.ks_name(), log_name(s.cf_name()));
+    b.set_default_time_to_live(gc_clock::duration{86400}); // 24h
+    b.set_comment(sprint("CDC log for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("stream_id", uuid_type, column_kind::partition_key);
+    b.with_column("time", timeuuid_type, column_kind::clustering_key);
+    b.with_column("batch_seq_no", int32_type, column_kind::clustering_key);
+    b.with_column("operation", int32_type);
+    b.with_column("ttl", long_type);
+    auto add_columns = [&] (const auto& columns) {
+        for (const auto& column : columns) {
+            b.with_column("_" + column.name(), column.type);
+        }
+    };
+    add_columns(s.partition_key_columns());
+    add_columns(s.clustering_key_columns());
+    add_columns(s.static_columns());
+    add_columns(s.regular_columns());
+    auto& mm = service::get_local_migration_manager();
+    return mm.announce_new_column_family(b.build(), false);
+}
+
+static future<> setup_desc(const schema& s) {
+    schema_builder b(s.ks_name(), desc_name(s.cf_name()));
+    b.set_comment(sprint("CDC description for %s.%s", s.ks_name(), s.cf_name()));
+    b.with_column("node_ip", inet_addr_type, column_kind::partition_key);
+    b.with_column("shard_id", int32_type, column_kind::partition_key);
+    b.with_column("created_at", timestamp_type, column_kind::clustering_key);
+    b.with_column("stream_id", uuid_type);
+    auto& mm = service::get_local_migration_manager();
+    return mm.announce_new_column_family(b.build(), false);
+}
+
+template <typename Func>
+static void for_each_endpoint(Func&& f) {
+    auto& topology = service::get_local_storage_service().get_token_metadata().get_topology();
+    for (const auto& dc : topology.get_datacenter_endpoints()) {
+        for (const auto& endpoint : dc.second) {
+            f(endpoint);
+        }
+    }
+}
+
+template <typename Func>
+static void for_each_shard(const gms::inet_address& endpoint, Func&& f) {
+    auto& snitch_ptr = locator::i_endpoint_snitch::get_local_snitch_ptr();
+    const unsigned int shard_count = snitch_ptr->get_shard_count(endpoint);
+    for (unsigned int shard_id = 0; shard_id < shard_count; ++shard_id) {
+        f(shard_id, shard_count);
+    }
+}
+
+static future<>
+populate_desc(service::storage_proxy& proxy, const schema& s) {
```

Comment:
This function makes some important assumptions about the state of DB when it's called, for example that `setup_desc` with the same schema `s` was previously called (and no `setup_desc`s with different schemas after that). It would make it easier to understand the function if these assumptions were written in a comment.

Understanding these assumptions will become much more important when we'll implement schema/topology changes.